File: services/engine/app/marketplace/slots.py
"""Booking slot rules and the Africa/Nairobi presenter (Sprint 004, spec #52 §HTTP API, #57).

Pure functions, imported by the create and counter endpoints so every proposal is validated the
same way and every BookingOut renders `proposedStartLocal` the same way. Times are stored in UTC
and the zone is applied here (D-16). The rules the Operator chose: any date inside the builder's
confirmed availability ranges, any start on the 30-minute grid from 08:00 to 18:00 Africa/Nairobi
(18:00 is the last start), duration 30 or 45 minutes, no buffer. Stitch's four fixed slots and
"30-min buffer" are not adopted (D-36 substitution). This module imports nothing from the API or
the store.
"""

from __future__ import annotations

from collections.abc import Iterable
from datetime import date, datetime, time, timedelta, timezone

NAIROBI = timezone(timedelta(hours=3), name="Africa/Nairobi")
GRID_MINUTES = 30
FIRST_START = time(8, 0)
LAST_START = time(18, 0)
DURATIONS = (30, 45)

OUTSIDE_AVAILABILITY = (
    "start is outside the builder's confirmed availability ({day} is in no range)"
)
OFF_GRID = "start must be on the 30-minute grid between 08:00 and 18:00 Africa/Nairobi"
BAD_DURATION = "duration must be 30 or 45 minutes"


def _require_aware(value: datetime, what: str) -> None:
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError(f"{what} must be an aware datetime")


def to_nairobi(value: datetime) -> str:
    """An aware datetime as an ISO 8601 string in Africa/Nairobi (`+03:00`)."""
    _require_aware(value, "datetime")
    return value.astimezone(NAIROBI).replace(microsecond=0).isoformat()


def valid_slot(
    start: datetime, duration_min: int, availability: Iterable[tuple[date, date]]
) -> str | None:
    """None when the proposal follows every rule, else the reason the API answers with 422."""
    _require_aware(start, "start")
    local = start.astimezone(NAIROBI)
    day = local.date()
    if not any(first <= day <= last for first, last in availability):
        return OUTSIDE_AVAILABILITY.format(day=day.isoformat())
    on_grid = (
        local.second == 0
        and local.microsecond == 0
        and local.minute % GRID_MINUTES == 0
        and FIRST_START <= local.time() <= LAST_START
    )
    if not on_grid:
        return OFF_GRID
    if duration_min not in DURATIONS:
        return BAD_DURATION
    return None

File: apps/web/src/lib/slots.ts
/**
 * The booking slot grid on the client (spec #52, Operator's slot rules): any date inside the
 * builder's confirmed availability ranges, any start on the 30-minute grid from 08:00 to 18:00
 * Africa/Nairobi (18:00 is the last start), duration 30 or 45 minutes. The engine enforces the
 * same rules (`slots.py`); the browser only offers what can pass. A proposal is sent as an ISO
 * string carrying the +03:00 offset, so no zone arithmetic happens here (D-16).
 */
import type { AvailabilityRange } from "@venture-route/contracts";

import { iso } from "./dates";
import { formatNairobi } from "./nairobi";

export const DURATIONS = [30, 45] as const;
export type Duration = (typeof DURATIONS)[number];

/** "08:00" … "18:00", every 30 minutes: 21 starts. */
export const GRID_STARTS: readonly string[] = Array.from({ length: 21 }, (_, index) => {
  const minutes = 8 * 60 + index * 30;
  return `${String(Math.floor(minutes / 60)).padStart(2, "0")}:${String(minutes % 60).padStart(2, "0")}`;
});

/** True when the calendar day lies inside one of the builder's confirmed ranges. */
export function inAvailability(day: Date, ranges: readonly AvailabilityRange[]): boolean {
  const date = iso(day);
  return ranges.some((range) => range.start <= date && date <= range.end);
}

/** `2026-09-24` + `10:30` → `2026-09-24T10:30:00+03:00`, the wire form of a proposal. */
export function localIso(day: Date, start: string): string {
  return `${iso(day)}T${start}:00+03:00`;
}

/** `10:30` + 45 → `11:15`. */
export function endOf(start: string, durationMin: Duration): string {
  const [h, m] = start.split(":").map(Number);
  const total = (h ?? 0) * 60 + (m ?? 0) + durationMin;
  return `${String(Math.floor(total / 60)).padStart(2, "0")}:${String(total % 60).padStart(2, "0")}`;
}

/** What the slot picker holds before a proposal is sent. */
export type SlotDraft = { day: Date | undefined; start: string | null; durationMin: Duration; note: string };

export const EMPTY_SLOT: SlotDraft = { day: undefined, start: null, durationMin: 30, note: "" };

/** "Thu 24 Sep 2026 · 10:30 – 11:15 EAT" for the summary, or null until day and start are chosen. */
export function summaryOf(draft: SlotDraft): string | null {
  if (!draft.day || !draft.start) return null;
  const local = formatNairobi(localIso(draft.day, draft.start));
  return `${local.replace(" EAT", "")} – ${endOf(draft.start, draft.durationMin)} EAT`;
}
